File: 2020/d1.py
""" https://adventofcode.com/2020/day/1 """

from typing import List


def part1(data: List[int]) -> int:
    """ O(n) solution """

    data_schema = [False] * 2020
    for e1 in data:
        e2 = 2020 - e1
        if e2 > 0 and data_schema[e2]:
            return e1 * e2
        else:
            data_schema[e1] = True

    return -1


def part2(data: List[int]) -> int:
    """ O(n^2) solution """

    data_schema = [False] * 2020
    for e1 in data:
        for e2 in data:
            e3 = 2020 - e2 - e1
            if e3 > 0 and data_schema[e3]:
                return e1 * e2 * e3
            else:
                data_schema[e1], data_schema[e2] = True, True

    return -1


if __name__ == "__main__":
    TEST = [int(line.strip()) for line in open("tests/d1.txt", "r")]
    PUZZLE = [int(line.strip()) for line in open("puzzles/d1.txt", "r")]

    assert part1(TEST) == 514579
    assert part2(TEST) == 241861950

    print(f"Part 1: {part1(PUZZLE)}")
    print(f"Part 2: {part2(PUZZLE)}")

File: 2021/src/aoc_2021/d1.clj
(ns aoc-2021.d1 (:require
                 [clojure.string :as str]))

(defn- parse-input [path]
  (->> path
       (slurp)
       (str/split-lines)
       (map #(Integer/parseInt %))))

(defn- count-increases [coll]
  (->> coll
       (partition 2 1)
       (filter (partial apply <))
       (count)))

(defn part-1
  "How many measurements are larger than the previous measurement?"
  [measurements]
  (count-increases measurements))

(defn part-2
  "How many sums are larger than the previous sum?"
  [measurements]
  (->> measurements
       (partition 3 1)
       (map (partial apply +))
       (count-increases)))

;; Puzzle: https://adventofcode.com/2021/day/1

(def test-input (parse-input "resources/tests/d1.txt"))
(def input (parse-input "resources/puzzles/d1.txt"))

(= (part-1 test-input) 7)
(time (part-1 input))

(= (part-2 test-input) 5)
(time (part-2 input))
